### DISCUSSION TITLE
Config.cfg Execution Logic

### DISCUSSION THREAD

**Discussion by koayst** [explosion/spaCy#9355] [Help: Coding & Implementations] [answered]
Hi,

I am trying to understand the logical flow on how Spacy uses config.cfg to create a model by studying [rel_componnent](https://github.com/explosion/projects/tree/v3/tutorials/rel_component).  I also did the tutorial found in [thinc.ai](https://github.com/explosion/thinc/blob/master/examples/00_intro_to_thinc.ipynb).

Based on this [config](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/configs/rel_tok2vec.cfg), how does Spacy knows it needs to build the model that is labelled as "rel_model.v1" ?

From thinc.ai tutorial, there is a common pattern (as shown below) after the config file is read and resolved.
```
  model = loaded_config ["model"]
  optimizer = loaded_config ["optimizer"]
  n_iter = loaded_config ["training"]["n_iter"]
  batch_size = loaded_config ["training"]["batch_size"]
  train_model(model, optimizer, n_iter, batch_size)
```

Can't seem to find similar pattern in the ["rel_components"](https://github.com/explosion/projects/tree/v3/tutorials/rel_component) files.

Could you please help me to understand the flow? I think I have looked every documentation I can read.  Something still does not connect.

Thanks.

**Comment by koayst** [comment]
In the project.yml file, under the "commands" section for "train_cpu", there is an option "-c"

script:
      - "python -m spacy train ${vars.tok2vec_config} --output training --paths.train ${vars.train_file} --paths.dev ${vars.dev_file} -c ./scripts/custom_functions.py"

Wonder what is the significance of "custom_functions.py" when it comes to training the model?

**Comment by thomashacker** [comment, marked as answer]
Hey koayst ✨,
I think, the missing piece to better understand how the config system works is the [Registry](https://thinc.ai/docs/api-config#registry), where you can map strings like `rel_model.v1` to a function which, in this case, would return a `thinc model`. You can also assign additional parameters when creating new registries. Here's the example from the REL code:

```
@spacy.registry.architectures("rel_model.v1")
def create_relation_model(
    create_instance_tensor: Model[List[Doc], Floats2d],
    classification_layer: Model[Floats2d, Floats2d],
) -> Model[List[Doc], Floats2d]:
    with Model.define_operators({">>": chain}):
        model = create_instance_tensor >> classification_layer
        model.attrs["get_instances"] = create_instance_tensor.attrs["get_instances"]
    return model
```

The `-c` parameter, when using `spacy train`, is important to resolve custom registry strings in your config, such as `rel_model.v1`.
You can read more about the different parameters in this [train section](https://spacy.io/api/cli#train)
